EventGenerator/src/CaloCalibGun_module.cc

```diff
@@ -0,0 +1,284 @@
+    {
+      art::ServiceHandle<art::TFileService> tfs;
+      _Ntupe  = tfs->make<TTree>("calibGun", "calibGun");
+      _Ntupe -> Branch("genErg", &_genErg, "genErg/F");
+      _Ntupe -> Branch("genTime", &_genTime, "genTime/F");
+      _Ntupe -> Branch("genCos", &_genCos, "genCos/F");
+      _Ntupe -> Branch("genPosX", &_genPosX, "genPosX/F");
+      _Ntupe -> Branch("genPosY", &_genPosY, "genPosY/F");
+      _Ntupe -> Branch("genPosZ", &_genPosZ, "genPosZ/F");
+    }
+    if (_mean < 0) throw cet::exception("RANGE") << "CaloCalibGun.mean must be non-negative "<< '\n';
+  }
+
+  //================================================================
+  void CaloCalibGun::produce(art::Event& event) {
+    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
+    PrimaryParticle primaryParticles;
+    MCTrajectoryCollection mctc;
+
+    _cal =  &*GeomHandle<mu2e::DiskCalorimeter>();
+    //_nPipes          = _cal->caloInfo().getInt("nPipes");
+    _pipeRadius      = _cal->caloInfo().getDouble("pipeRadius");
+    _pipeTorRadius   = _cal->caloInfo().getVDouble("pipeTorRadius");
+    _randomRad       = _cal->caloInfo().getVDouble("pipeTorRadius");
+    _zPipeCenter     = _cal->disk(1).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(1).geomInfo().size().z()/2.0-_pipeRadius);//disk =1
+    // we normalize to the volume of the pipe (proportional to 2*pi*R if they have all the same radius) to draw a random number from which to generate the photons TODO - is this used?
+    double sumR(0);
+    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {sumR+=d; d = sumR; });
+    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {d /= sumR;});
+    
+    unsigned int nGen = _mean;
+    //Define the parameters of the pipes:
+    
+    // angle of large torus in degrees
+    double phi_lbd[5] = {161.34, 149.50, 139.50, 132.07, 125.39};
+    // angle of small torus in degrees
+    double phi_sbd[5] = {84.63, 85.28, 85.79, 86.20, 86.53};
+    // angle of the end point
+    double phi_end[5] = {3.96, 10.53, 15.80, 20.16, 23.84};
+    // center position y of the small torus
+    double ysmall[5] = {432.2, 480.5, 524.3, 564.7, 602.5};
+    // radius of small torus
+    double radSmTor = 41.0;
+    // first center position x of the small torus
+    double xsmall = 71.0;
+    // distance of the small torus center
+    double xdistance = 60.0;
+    // inner radius of the manifold
+    double rInnerManifold = 681.6; // 713.35 mm - 1.25 in (31.75 mm)
+    vector<float> sign{-1.0, 1.0};
+
+    for (unsigned int ig = 0; ig < nGen; ++ig) 
+    {
+      // only in the front of 2nd disk.
+      //int nDisk        = 1;
+      //_zPipeCenter     = _cal->disk(nDisk).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(nDisk).geomInfo().size().z()/2.0-_pipeRadius);
+
+      double xpipe, ypipe, zpipe;
+      //Pick position
+      int xsn = rint(_randFlat.fire());
+      int ysn = rint(_randFlat.fire());
```

Comment:
As I recall, the default range for randFlat is a [0,1) or [0,1].  So xsn and ysn will always be 0.   Is this the intended behaviour?

Reply:
Yes, it is intended to select a random sign either 0,1 from the sign vector

Reply:
Sorry.  I misread it as int() not rint().  However this code is dangerous since fegetround is global that can be modified by code unrelated to yours.  This is a legacy of it's 

   https://www.cplusplus.com/reference/cmath/rint/
   https://www.cplusplus.com/reference/cfenv/fegetround/

It would be better not to rely on constructs whose behaviour can be changed by code in other places.  I believe that 
int xsn = round(x); will do this more safely.

Reply:
Thanks - will change